File: prob51.hs


isPrime :: Integer -> Bool
isPrime x 
	| x <= 1 	= False
	| otherwise = isPrime' x 2
	where isPrime' num div = if div * div > num
							 then True
							 else (num `mod` div /= 0) && isPrime' num (div+1)


replaceDigit :: String -> Char -> Char -> String
replaceDigit [] _ _ = error "empty number"
replaceDigit all@(x:xs) old new 
	| x == old && new == '0' = "0"
	| otherwise = map (\dig -> if dig == old then new else dig) all


genSequence :: String -> Char -> [Integer]
genSequence num old
	| old `elem` num = filter (\x -> isPrime x) (map (\x -> read x) replaced)
	| otherwise = []
	where replaced = map (\x -> replaceDigit num old x) ['0'..'9']


testPrimeFam :: Integer -> Int -> Bool
testPrimeFam x len 
	| length cand < len = False
	| otherwise = any (\lst -> (length lst) >= len) (map (\old -> genSequence num old) cand)
	where
		num = show x
		cand = [minimum num..'9']


searchFam :: Int -> Integer
searchFam len = searchFam' l len
				where l = [y | y <- [2..], isPrime y]


searchFam' :: [Integer] -> Int -> Integer
searchFam' (x:xs) len = if testPrimeFam x len
						then x
						else searchFam' xs len

File: prob51.py
from math import sqrt
import sys

def isPrime (x):
	if x <= 1:
		return False		  
	for div in xrange(2, int(sqrt(x))):
		if x % div == 0:
			return False
	return True


def minDigits (num):
	return min (str(num))


def replaceNum (num, old, new):
	if old not in str(num):
		return 0
	elif str(num).startswith(old) and new == '0':
		return 0
	else:
		return int(''.join(map(lambda x : new if x == old else x, str(num))))


if __name__ == "__main__":
	num = 56004
	digs = set(['0', '1', '2'])
	char = '0123456789'

	while True:
		if isPrime (num) and minDigits (num) in digs:
			for dig in range(int(minDigits(num)), 9):
				temp = filter (lambda x : isPrime(x), map (lambda x : replaceNum(num, str(dig), x), char) )
				if len (temp) == 8: 
					print num
					print temp
					sys.exit (0)
		num += 1

